### PR TITLE
web: behaviour: fix scroll on build history list

### LINKED ISSUE #5383: Scrolling build numbers list is broken
## Bug Report

Scrolling through the list of builds does not work - it always jumps to the most recent build

## Steps to Reproduce

1. Open a job with many builds
2. Use your mouse wheel to scroll

## Expected Results

I can scroll to older builds

## Actual Results

It fights me, and jumps back to the start of the list

## Version Info

* Concourse version: 6.0
* Deployment type (BOSH/Docker/binary): N/A
* Infrastructure/IaaS: N/A
* Browser (if applicable): Chrome
* Did this used to work? Yes

### PR DESCRIPTION
**Pull request by aoldershaw**
# Existing Issue

Fixes #5383 .

# Changes proposed in this pull request

* Introduce `scrollCoordsSimple` function for dealing with simple cases of computing scroll position (that don't require knowing the coordinates of the source and parent elements)
* Replace the deprecated DOM event `mousewheel` with `wheel`, which fixes scrolling on Firefox (which seems to have been broken for a while)

Note that this behaviour could be shoehorned in to the more complicated `scrollCoords` function, but there is a noticeable performance hit in scrolling due to the increased number of calls out to `Elm.Browser`.

# Contributor Checklist
- [x] Updated release notes (located at https://github.com/concourse/concourse/tree/master/release-notes)

# Reviewer Checklist
- [ ] Code reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed

### CODE REVIEW

**Comment by cirocosta** [comment]
Hey @aoldershaw ,

I was able to get scrolling working on Chrome, but not on Firefox :thinking:   - do you think it'd be hard to adapt for it too? the fix you provided doesn't seem to deal with anything that's platform dependent, which makes me wonder if it even worked before on Firefox at all :thinking:  

thanks!

**Comment by aoldershaw** [comment]
Hey @cirocosta, thanks for testing in multiple browsers! Weird that it doesn't work in Firefox - this is essentially just reverting the scrolling behaviour to 495904da500188d89a9c9e62071d4ac0e3205ed7.

It also seems to not be the same scrolling issue as I set out to fix - on Chrome, you could scroll horizontally for a split second before having it forced back to the latest build, whereas on Firefox, it doesn't seem to move at all.

I'll take a look!

**Comment by aoldershaw** [comment]
@cirocosta turns out we were using a deprecated event `onmousewheel` when we should have been using `onwheel` - https://developer.mozilla.org/en-US/docs/Web/API/Element/wheel_event.

Seems to work on Chrome and Firefox now

**Comment by jomsie** [comment]
scrolling looks good on both chrome/safari (on osx), but firefox (on osx) is very slow scrolling... takes a whole scroll to go back 1 or 1.5 builds. (might be a firefox issue?)

still, slow scrolling beats no scrolling at all!

**Comment by aoldershaw** [comment]
> scrolling looks good on both chrome/safari (on osx), but firefox (on osx) is very slow scrolling... takes a whole scroll to go back 1 or 1.5 builds. (might be a firefox issue?)

Huh, weird...I hadn't noticed since I was just using my trackpad, but I am experiencing the same with a mouse. The `deltaY` values seem to be on completely different scales between Chrome and Firefox...not sure how best to deal with that

**Comment by jomsie** [comment]
> > scrolling looks good on both chrome/safari (on osx), but firefox (on osx) is very slow scrolling... takes a whole scroll to go back 1 or 1.5 builds. (might be a firefox issue?)
> 
> Huh, weird...I hadn't noticed since I was just using my trackpad, but I am experiencing the same with a mouse. The `deltaY` values seem to be on completely different scales between Chrome and Firefox...not sure how best to deal with that. We could try just checking the sign of `deltaY` and scale it to a fixed value, but that could be annoying trying to scroll a small amount.
> 
> EDIT: ah, it may be the case that the `deltaMode` differs (Firefox is measuring in terms of lines, rather than pixels - https://stackoverflow.com/questions/20110224/what-is-the-height-of-a-line-in-a-wheel-event-deltamode-dom-delta-line). I'll take a look

👀 look tomorrow

**Comment by aoldershaw** [comment]
Okay @jomsie, I've added some handling for `deltaMode`, which makes it feel more natural on Firefox. Let me know what you think!

**Comment by jomsie** [comment]
i'll review this after lunch!!!

File: web/elm/src/Message/Effects.elm [hunk 1]
```diff
@@ -657,19 +657,19 @@ scroll : ScrollDirection -> String -> Cmd Callback
...
-            scrollCoords id id (always 0) (.srcElem >> .viewport >> .y >> (+) -60)
+            scrollCoordsSimple id (always 0) (.viewport >> .y >> (+) -60)
 
         ToBottom ->
-            scrollCoords id id (always 0) (.parentElem >> .scene >> .height)
+            scrollCoordsSimple id (always 0) (.scene >> .height)
 
         Sideways delta ->
-            scrollCoords id id (.srcElem >> .viewport >> .x >> (+) -delta) (always 0)
+            scrollCoordsSimple id (.viewport >> .x >> (+) -delta) (always 0)
```
Comment (jomsie) [inline web/elm/src/Message/Effects.elm:672]: the things firefox makes you do 😆

File: web/elm/src/StrictEvents.elm [hunk 2]
```diff
@@ -28,6 +30,12 @@ type alias ScrollState =
     }
 
 
+type DeltaMode
```
Comment (jomsie) [inline web/elm/src/StrictEvents.elm:33]: the things firefox makes you do 😆

**Comment by jomsie** [comment]
Just some evidence! 

safari:
safari

chrome:
chrome

firefox:
firefox

File: web/elm/src/Build/Header/Header.elm [hunk 3]
```diff
@@ -394,12 +395,23 @@ update msg ( model, effects ) =
     case msg of
         ScrollBuilds event ->
             let
+                scrollFactor =
+                    case event.deltaMode of
+                        DeltaModePixel ->
```
Comment (jomsie) [inline web/elm/src/Build/Header/Header.elm:400]: Just a side comment, not super important, how'd you get the 20 / 800 for line / page?

Comment (aoldershaw) [reply web/elm/src/Build/Header/Header.elm:400]: Totally fair question! Tried to give some justification for it in the commit message I added that in.

tl;dr 20 for Line is based on the font-size of the scroll bar. The 800 for Page is totally arbitrary - I took it from some Facebook library. I haven’t been able to produce a DeltaModePage event, so not sure if it feels natural, but it’s better than nothing!
